src/circuit/boolean.rs

```diff
@@ -1115,4 +1412,171 @@ mod test {
+                    let a;
+                    let b;
+                    let c;
+
+                    // ch = (a and b) xor ((not a) and c)
+                    let expected = (first_operand.val() & second_operand.val()) ^
+                                   ((!first_operand.val()) & third_operand.val());
+
+                    {
+                        let mut dyn_construct = |operand, name| {
+                            let cs = cs.namespace(|| name);
+
+                            match operand {
+                                OperandType::True => Boolean::constant(true),
+                                OperandType::False => Boolean::constant(false),
+                                OperandType::AllocatedTrue => Boolean::from(AllocatedBit::alloc(cs, Some(true)).unwrap()),
+                                OperandType::AllocatedFalse => Boolean::from(AllocatedBit::alloc(cs, Some(false)).unwrap()),
+                                OperandType::NegatedAllocatedTrue => Boolean::from(AllocatedBit::alloc(cs, Some(true)).unwrap()).not(),
+                                OperandType::NegatedAllocatedFalse => Boolean::from(AllocatedBit::alloc(cs, Some(false)).unwrap()).not(),
+                            }
+                        };
+
+                        a = dyn_construct(first_operand, "a");
+                        b = dyn_construct(second_operand, "b");
+                        c = dyn_construct(third_operand, "c");
+                    }
+
+                    let maj = Boolean::sha256_ch(&mut cs, &a, &b, &c).unwrap();
+
+                    assert!(cs.is_satisfied());
+
+                    assert_eq!(maj.get_value().unwrap(), expected);
+
+                    if first_operand.is_constant() ||
+                       second_operand.is_constant() ||
+                       third_operand.is_constant()
+                    {
+                        if first_operand.is_constant() &&
+                           second_operand.is_constant() &&
+                           third_operand.is_constant()
+                        {
+                            assert_eq!(cs.num_constraints(), 0);
+                        }
+                    }
+                    else
+                    {
+                        assert_eq!(cs.get("ch"), {
+                            if expected {
+                                Fr::one()
+                            } else {
+                                Fr::zero()
+                            }
+                        });
+                        cs.set("ch", {
+                            if expected {
+                                Fr::zero()
+                            } else {
+                                Fr::one()
+                            }
+                        });
+                        assert_eq!(cs.which_is_unsatisfied().unwrap(), "ch computation");
+                    }
+                }
+            }
+        }
+    }
+
+    #[test]
+    fn test_boolean_sha256_maj() {
```

Comment:
Refactor this and ``test_boolean_sha256_ch`` using a helper that tests any boolean × boolean → boolean function given a closure specifying the function.